### DISCUSSION TITLE
About the realization of Qos 2

### DISCUSSION THREAD

**Discussion by StanLi-ops** [ntex-rs/ntex-mqtt#85] [Q&A] [answered]
Whether the implementation of QoS 2 is currently considered. Or can it be achieved by calling send_at_least_once()?

**Comment by fafhrd91** [comment, marked as answer]
I don't need qos2, so I never tried to implement it. but PR is welcome.

**Comment by StanLi-ops** [comment]
Hello @fafhrd91  
I am going to implement QoS2 for a project. Currently I am learning the relevant code of QoS1 in the project.
In send_at_least_once() I didn't find any related operation about "DUP".

**Comment by fafhrd91** [comment]
what do you mean "dup" related operation? 

https://github.com/ntex-rs/ntex-mqtt/blob/09d49308eb66f5c10ade2ecb3dc4ea5623e8665f/src/v3/sink.rs#L175

**Reply by StanLi-ops** [reply]
> 

1) Sender sends a PUBLISH packet with message data to Receiver, and saves this PUBLISH packet locally.

2) After receiving the PUBLISH packet, the Receiver sends a PUBACK packet to the Sender. The PUBACK packet has no message body (Payload), and has a Packet Identifier in the Variable header, and it receives The Packet Identifier in the PUBLISH package is the same.

3) After receiving the PUBACK, the Sender finds the locally saved PUBLISH packet according to the Packet Identifier in the PUBACK packet, and then discards it, and the sending of a message is completed.

4) If the Sender does not receive the PUBACK corresponding to the PUBLISH packet within a period of time, it sets the DUP flag of the PUBLISH packet to 1 (representing a resent PUBLISH packet), and then resends the PUBLISH packet. Repeat this process until a PUBACK is received, then go to step 3.

Will the "step 4" be implemented by the framework? Or is it implemented by the user himself?

**Reply by fafhrd91** [reply]
step 4 needs to be implemented by user

**Reply by StanLi-ops** [reply]
> 第 4 步需要用户执行

Can you provide a complete (including step 4) example of QoS 1?

**Reply by fafhrd91** [reply]
this requires time, which i dont have at the moment. it may be included to the framework but as separate method

**Reply by fafhrd91** [reply]
let me think, this seems more complicated than i thought

**Reply by StanLi-ops** [reply]
> 

I tried to implement "step 4" by calling "send_at_least_once()", but all failed.

**Reply by StanLi-ops** [reply]
> let me think, this seems more complicated than i thought

This is the qos1 I'm currently trying to implement, I'm not sure it's the right way to do it.

```Rust
async fn publish_v5(
    session: v5::Session<MySession>,
    publish: v5::Publish,
) -> Result<v5::PublishAck, MyServerError> {
    log::info!(
        "incoming publish ({:?}) : {:?} -> {:?}",
        session.state(),
        publish.id(),
        publish.topic()
    );

    let sink = session.sink();

    let builder = sink.publish("test", Bytes::from("test"));
    let interval = ntex::time::interval(core::time::Duration::from_secs(1));

    let send_at_least_once = futures::FutureExt::fuse(builder.send_at_least_once());
    let timeout = futures::FutureExt::fuse(interval.tick());

    futures::pin_mut!(send_at_least_once, timeout);

    loop {
        futures::select! {
           _ = send_at_least_once  => {
               println!("send_at_least_once");
               break
            },
            _ = timeout => println!("timeout"),
        }
    }

    Ok(publish.ack())
}
```

**Reply by fafhrd91** [reply]
problem is in ntex-mqtt, in case of timeout ack handler wont be removed so could stall or fail for different ack. ntex needs some changes

**Reply by StanLi-ops** [reply]
> problem is in ntex-mqtt, in case of timeout ack handler wont be removed so could stall or fail for different ack. ntex needs some changes

qos2 has two retransmission mechanisms similar to qos1, so in order to ensure the same usage as qos1 as much as possible, I hope this problem can be solved as soon as possible.

**Reply by StanLi-ops** [reply]
> problem is in ntex-mqtt, in case of timeout ack handler wont be removed so could stall or fail for different ack. ntex needs some changes

I read the code of v5::sink::send_at_least_once_inner(). My understanding is that each publish has a corresponding "publish ack channel" to process the returned ack, so I don't quite understand the meaning of the sentence "could stall or fail for different ack". Can you explain in detail?
